test/results-shape.js

```diff
@@ -0,0 +1,164 @@
+        for (const subObject of subObjects) {
+            for (const subKey of Object.keys(subObject)) {
+                //recursively look down the tree
+                keyAndChildrenAreValid(
+                    subObject,
+                    subKey,
+                    shapeObject[key].subObject
+                );
+            }
+        }
+    }
+}
+
+function requiredKeyAndChildrenArePresent(object, shapeObject, key) {
+    //verify the key is legal
+    if (shapeObject[key].required) {
+        expect(key in object, `${key} is required but is missing!`).to.exist;
+    }
+    if (object[key]) {
+        expect(
+            shapeObject[key].type,
+            `${key} is of unexpected type ${typeof object[key]}!`
+        ).to.include.members([typeof object[key]]);
+    }
+    if (shapeObject[key].subObject && object[key]) {
+        //if there is an object within this key, evaluate it
+        const subObjectKeys = Object.keys(shapeObject[key].subObject);
+        for (const subObjectKey of subObjectKeys) {
+            //recursively look down the tree
+            requiredKeyAndChildrenArePresent(
+                object[key],
+                shapeObject[key].subObject,
+                subObjectKey
+            );
+        }
+    }
+}
+
+describe("shape of return object", async function () {
+    let finalNumberTests;
+    before(async function () {
+        const originalConsole = console.log;
+        this.timeout(10 * 60 * 1000); //10 minutes, shouldn't take that long though
+        console.log = () => {};
+        sinon
+            .stub(getGeocode, "getAllCoordinates")
+            .callsFake((locations, ...rest) => {
+                return locations.map((location) => ({
+                    ...location,
+                    latitude: 100,
+                    longitude: -100,
+                }));
+            });
+        results = await require("../scraper")
+            .handler()
+            .then((res) => res.results);
+        const results_no_browser = await require("../scrapers_no_browser")
+            .handler()
+            .then((res) => res.results);
+        results = [...results, ...results_no_browser];
+        finalNumberTests = results.length;
+        console.log = originalConsole;
+    });
+
+    it("has no unauthorized keys", (done) => {
+        let testNumber = 0;
+        for (const result of results) {
+            console.log(result);
+            for (const key of Object.keys(result)) {
+                keyAndChildrenAreValid(result, key, keys);
+            }
+            testNumber++;
+        }
+
+        if (testNumber === finalNumberTests) {
+            done();
+        } else {
+            throw new Error(
+                `${testNumber} tests run, expected ${finalNumberTests}`
+            );
+        }
+    });
+
+    it("isn't missing any required keys", (done) => {
+        let testNumber = 0;
+        for (const result of results) {
+            console.log(result);
```

Comment:
and... this is producing a lot of output during the test

Reply:
Removed the console.logs and added a JSON.stringify when a test fails to help with context (that's why I added the console.logs originally - there was no object showing me what the heck was wrong and where)